### ISSUE TITLE
Why does Lasagne use a "Lazy" construction architecture ?

### ISSUE THREAD

**Issue by unrealwill** [Lasagne/Lasagne#623]
Lasagne doesn't work as I thought it would. 
The network is generated during the call to lasagne.layers.get_output().
Instinctively I would have constructed the layer theano symbolic output upon call of the layer constructor. 
What is the main use case for the lazy construction of the network ?

My main issues with regards to the lazy approach Lasagne use : 
-requirement to only do one get_output to share the same random numbers 
-It is that it's hard to have multiple outputs, it's hard to have multiple inputs, it is hard to do some things which should be one-liners (like an elemwise product layer). 
-There is also the problem that it's hard to reuse existing layers to create new layers (for example to create a "highway network" (a simple operation on the result from 2 dense networks) it would have been useful to reuse the dense networks. 
-Harder to debug because it sometime crashes on the get_output line for an error which happens earlier.
-Lazy evaluation usually needs harder treatment to avoid things being computed twice (a proper graph construction and traversal which is what theano already do so why redo it again in Lasagne)
-I can't yet thought of a good use case for which the lazy construction of the network would be better than it's immediate construction.

**Comment by f0k** [comment]
> Instinctively I would have constructed the layer theano symbolic output upon call of the layer constructor.
> What is the main use case for the lazy construction of the network ?

The `get_output()` call accepts keyword arguments that can modify the behaviour of the network. The first use case for this was to allow defining a single network architecture to be used with dropout (for training) and without dropout (for testing). For this, we cannot construct the output incrementally on constructing the network, as we need to generate two different expressions later.

If we restricted the API such that every layer creates exactly one expression for training and one expression for testing, we could still construct these two expressions incrementally, but it wouldn't be possible to create even more variants.

> requirement to only do one get_output to share the same random numbers 

See #104 for a discussion of alternative approaches and why we chose this one.

> It is that it's hard to have multiple outputs, it's hard to have multiple inputs, it is hard to do some things which should be one-liners (like an elemwise product layer).

Taking care to have a single `get_output()` call makes multiple-output networks a bit more involved, but what problems do you see for multiple-input networks?

> Harder to debug because it sometime crashes on the get_output line for an error which happens earlier.

This is indeed annoying. We try to catch such errors in the constructor (this usually gives a better description of what went wrong than just creating the symbolic expression and letting Theano fail). If you find errors that are caught too late, we're happy to improve things.

----

Note that you're free to use Lasagne with immediate construction:
```python
input_var = ...
network = InputLayer(input_shape, input_var)
expr = input_var
network = Conv2DLayer(network, ...)
expr = network.get_output_for(expr)
network = Conv2DLayer(network, ...)
expr = network.get_output_for(expr)
network = DenseLayer(network, ...)
expr = network.get_output_for(expr)
```
If you prefer this, you can write a short helper function or a model class that combines adding a layer and updating the expression (or the two expressions if you want to use dropout or batch normalization). You'll still have the benefits of shape propagation, parameter initialization and parameter collection.

**Comment by unrealwill** [comment]
Thank you for your answer. I played a little with Lasagne today and I understand better now and hopefully won't get burned again by the get_output().
For the multiple inputs network I had the problem of not inheriting from MergeLayer.
